test/e2e/storage/in_tree_volumes.go

```diff
@@ -0,0 +1,451 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package storage
+
+import (
+	"fmt"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	"k8s.io/api/core/v1"
+	storagev1 "k8s.io/api/storage/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilerrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/rand"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/kubernetes/test/e2e/framework"
+	"k8s.io/kubernetes/test/e2e/storage/utils"
+)
+
+var testDrivers = map[string]func() testDriver{
+	"NFS":                        initNFSDriver,
+	"GlusterFS":                  initGlusterFSDriver,
+	"iSCSI [Feature:Volumes]":    initISCSIDriver,
+	"Ceph-RBD [Feature:Volumes]": initRbdDriver,
+	"CephFS [Feature:Volumes]":   initCephFSDriver,
+	"hostpath":                   initHostpathDriver,
+	"hostpathSymlink":            initHostpathSymlinkDriver,
+	"emptydir":                   initEmptydirDriver,
+	"cinder":                     initCinderDriver,
+	"gce":                        initGceDriver,
+	"vSphere":                    initVSphereDriver,
+	"azure":                      initAzureDriver,
+	"aws":                        initAwsDriver,
+}
+
+var testSuites = map[string]func() testSuite{
+	"volumes":                          initVolumesTestSuite,
+	"volumeIO":                         initVolumeIOTestSuite,
+	"volumeMode [Feature:BlockVolume]": initVolumeModeTestSuite,
+	"subPath":                          initSubPathTestSuite,
+}
+
+var _ = utils.SIGDescribe("In-tree Volumes", func() {
+	f := framework.NewDefaultFramework("volumes")
+
+	var (
+		cs     clientset.Interface
+		ns     *v1.Namespace
+		node   v1.Node
+		config framework.VolumeTestConfig
+
+		driver       testDriver
+		testSuite    testSuite
+		testResource testResource
+		testInput    testInput
+		testVolType  testVolType
+	)
+
+	BeforeEach(func() {
+		cs = f.ClientSet
+		ns = f.Namespace
+		nodes := framework.GetReadySchedulableNodesOrDie(cs)
+		node = nodes.Items[rand.Intn(len(nodes.Items))]
+		config = framework.VolumeTestConfig{
+			Namespace: ns.Name,
+			Prefix:    "volume",
+		}
+	})
+
+	for driverName, initDriver := range testDrivers {
+		curDriverName := driverName
+		curDriver := initDriver()
+
+		for testSuiteName, testSuiteInit := range testSuites {
+			curTestSuiteName := testSuiteName
+			curTestSuite := testSuiteInit()
+			testPatterns := curTestSuite.getTestPatterns()
+
+			for testPatternName, testPattern := range testPatterns {
+				curTestPatternName := testPatternName
+				curTestPattern := testPattern
+				testInput = initTestMetadata(curTestSuiteName, curTestPatternName)
+				curTestInput := testInput
+				testFunc := curTestSuite.getTestFunc(curTestPattern, curDriver)
```

Comment:
Fixed as suggested. 

Also, in the same commit, I add a code to skip deleting resource and driver when it it is not created by skip. I found that unnecessary PV deletion routine for other clouds is running and taking so much time. It will be worth skipping, although tests won't fail. 

PTAL